Configure

```diff
@@ -1326,13 +1330,14 @@ if ($strict_warnings)
 	my $wopt;
 	die "ERROR --strict-warnings requires gcc or gcc-alike"
             unless defined($predefined{__GNUC__});
-	foreach $wopt (split /\s+/, $gcc_devteam_warn)
+	foreach $wopt (split /\s+/, $devteam_common_warn)
 		{
 		$config{cflags} .= " $wopt" unless ($config{cflags} =~ /(?:^|\s)$wopt(?:\s|$)/)
 		}
-	if (defined($predefined{__clang__}))
+	foreach my $ecc (grep { defined($predefined{$_}) }
+			 keys %devteam_additional_warn)
```

Comment:
Formally speaking this is a bit too fuzzy to rely upon. For following reason. As we know, given compiler can pretend to have multiple "personalities". For example clang defines both `__GNUC__` and `__clang__`. Compiler that triggered this discussion, icc, also defines `__GNUC__` and something for itself, `__ICC`. In other words there is nothing that prevents compiler to have multiple personalities, and even more than two. And this ask for following question. Is order in which command line is assembled significant? Trouble is that if it is, then we have no good control over it, order in which `key %devteam_additional_warn` are listed. Moreover, can't order even change as you add more entries? Bottom line is that I find it hard to endorse this suggested way to handle warning flags... Because I can't see it working reliably in long run. I'd argue that it should be more deterministic [than suggested], so that we can have better control...

Reply:
As for being more deterministic. Once again I want to ask to imagine if there was picker(warning) thing. Imagine there is no check for "--strict-warnings requires gcc or gcc-alike". Instead it checks if there is presence of picker(warning)... Well, one can still argue that check for gcc-alike-ness is justified by compiler's ubiquity, so that it might make sense to check for picker(warning) *and* gcc-alike-ness. Though maybe with gcc-alike-ness taking precedence, so that picker(warning) would be a *suffix* to gcc warnings. As more practical example, consider there was contributed 50-icc.c with picker(warning) that shuts specific gcc warnings...

Reply:
> check for gcc-alike-ness is justified by compiler's ubiquity...

... and its striking personality. The pun is reference to the fact that "multiple-personality" compilers is somewhat common place, and the additional "personality" is likely to be gcc.

Reply:
Well, err, yeah.  We did have `$ecc` before, which gave the determinism for this, I'm adapting to the situation.  Are you suggesting I reinstate `$ecc` and the way its value was defined?

Mind you, there *is* a radically different way to approach this, and that's to have compiler configs, marked as templates, and the final configs can inherit from them.  That way, we'd only need to let `--strict-warnings` enable a flag in `%config`, the rest is just a story of what we put in each compiler config.

Reply:
... we still need to be able to identify individual compilers somehow, though

Reply:
I think we're thinking in the same direction.

Reply:
> Are you suggesting I reinstate $ecc

No.

> radically different way to approach this,

And picker(warning) is something along that line. However! Current check for "personality" has rather strong side. Thing is that you don't know if any given compiler has the additional personality. If might have been added in some specific release, [for example icc didn't have `__GNUC__` personality to start with]. Or you can't be sure about "primary" personality because you can't make an assumption about what's hiding behind /usr/bin/cc [for example as far as I understand FreeBSD changed from GNU C to clang in some release]. The question I wonder is if it makes sense to *combine* deterministic with heuristic. But I can't find the suggested approach to be the right answer... Do note that I'm not saying that it's "wrong answer", only that I *for one* can't see that it's right.

Reply:
In a way, I like the picker(warning) idea...  the only really icky thing is that the way the configs are currently structure, that means code duplication to every gcc and clang config, and you probably know by now how much I dislike that...

Reply:
As already said, how about combining them? I.e. how about leaving "personality" detection as is, and then adding picker(warning) *if* there is one. This way you don't need *any* picker(warning) for gcc or clang targets, and for others one can use this opportunity to mask unsuitable warnings and/or add compiler-specific. And for those compilers which don't have gcc "personality" picker(warning) would be just flags... Yeah, one can argue it's too vague and might be counter-intuitive, but let's consider alternatives: a) picker(warning) flags to each target; b) leave "personality" detection as is and instruct users to add desired [mask] flags on config command line. Of the two I'd say b). Or in other words I'd suggest to do nothing if a) is suggested.